saasherder/changelog.py

```diff
@@ -1,140 +1,186 @@
-        remote = service['url'].strip('/')
-        template = '[%h](_REMOTE_/commit/%H) %s'
+    def last_changed(self, service):
+        """Get the commit time for service at specific commit"""
+        return self.changelog.service_run(service, "git log -1 --format=%cd {}".format(service['new'])).strip()
 
-        cmd = ("cd {} && git log --format='{}' {}...{}"
-               .format(worktree, template, old, new))
+class Changelog(object):
 
-        return self.run(cmd).replace('_REMOTE_', remote)
+    workspace = "_workspace/"
 
-    def last_changed(self, service, commit):
-        """Get the commit time for service at specific commit"""
+    def __init__(self, parent):
+        self.parent = parent
 
-        worktree = self.worktree(service['name'])
-        cmd = "cd {} && git log -1 --format=%cd {}".format(worktree, commit)
+        # Key: url of the service
+        # Value: list of services that have changed
+        self.changed_services = []
 
-        return self.run(cmd).strip()
+    def fetch_diff(self, now, previous):
+        """Return the difference between 2 service lists
 
-    @staticmethod
-    def markdown(changelog, old, new):
-        """Pretty print the change log
+        This is not a commutative relationship. We are looking for services in
+        first list which differ from that in second list. a ⊕ b ≠ b ⊕ a.
 
-        The incoming object is a list of `(name, timestamp, messages)` tuples.
+        This gets messy when the hash refers to a branch or none. Ignoring such
+        services for now.
         """
 
-        url = "https://github.com/openshiftio/saas-openshiftio"
+        # All the services that changed. We will store only those that exist in
+        # now and in previous, for which 'hash' is defined in both
+        changed_services = {}
+        for service_name, service in now.items():
+            try:
+                prev_service = previous[service_name]
+                service_hash = service['hash']
+                prev_service_hash = prev_service['hash']
+            except KeyError:
+                # Skip if prev_service is not found
+                # or if 'hash' is not defined for either service
+                continue
+
+            if service_hash is None or prev_service_hash is None:
+                continue
+
+            url = service['url'].rstrip('/')
+
+            if prev_service_hash != service_hash:
+                if url in changed_services:
+                    changed_services[url]['names'].append(service_name)
+                else:
+                    diff_item = dict(name=service_name,
+                                     names=[service_name],
+                                     url=url,
+                                     new=service_hash,
+                                     old=prev_service_hash)
+
+                    changed_services[url] = diff_item
+
+        self.changed_services = changed_services.values()
```

Comment:
I'd avoid this line if I can - mutating variables in a different scope.

Reply:
I'm unsure what you mean? Can you further explain?